extensions/security/runtime/src/main/java/io/quarkus/security/runtime/interceptor/SecurityConstrainer.java

```diff
@@ -0,0 +1,164 @@
+package io.quarkus.security.runtime.interceptor;
+
+import static java.util.Arrays.asList;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import javax.annotation.security.DenyAll;
+import javax.annotation.security.PermitAll;
+import javax.annotation.security.RolesAllowed;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import io.quarkus.security.Authenticated;
+import io.quarkus.security.identity.SecurityIdentity;
+import io.quarkus.security.runtime.ForbiddenException;
+import io.quarkus.security.runtime.UnauthorizedException;
+
+/**
+ * @author Michal Szynkiewicz, michal.l.szynkiewicz@gmail.com
+ *         <br>
+ *         Date: 03/10/2019
+ */
+@Singleton
+public class SecurityConstrainer {
+
+    private static final List<Class<? extends Annotation>> SECURITY_ANNOTATIONS = asList(Authenticated.class, DenyAll.class,
+            PermitAll.class, RolesAllowed.class);
+
+    private final Map<Method, Optional<Check>> checkForMethod = new HashMap<>();
+
+    @Inject
+    SecurityIdentity identity;
+
+    public void checkRoles(Method method, Collection<Annotation> interceptorBindings) {
+        getCheck(method, interceptorBindings)
+                .ifPresent(check -> check.apply(identity));
+    }
+
+    private Optional<Check> getCheck(Method method, Collection<Annotation> interceptorBindings) {
+        return checkForMethod.computeIfAbsent(method, m -> determineSecurityCheck(m, interceptorBindings));
+    }
+
+    private Optional<Check> determineSecurityCheck(Method method, Collection<Annotation> interceptorBindings) {
```

Comment:
Is there any thought to leveraging build time information for the types of checks that are being performed here?

Reply:
that would definitely be good. 
OTOH, determining the check is done only once for method.

Reply:
+1 to doing all this at build time, but I am not sure if our current API's really support it. @mkouba any ideas about this? We really want method level interceptors to be able to override class level interceptors, and also some way of passing in information to the interceptors (i.e. the roles allowed). At the moment this is not really possible with CDI but it would be great if we could support it somehow.

Reply:
Normally you solve this simply by nature of interceptor resolution - by bindings. But the combinations here don't really allow that - there are 4 interceptors and they all share the same logic which only then decides what should be done.

> some way of passing in information to the interceptors (i.e. the roles allowed)

Not sure I follow? You can store anything inside `InvocationContext` and pass it around between the interceptors if that's what you mean.

As for method and class bindings, we could perhaps store them separately in arc and expose them as we now do the combined version? So at least user can see what was defined where. But otherwise I think interceptors spec already defined that method ones have a priority over class ones, so I suppose if you place the same annotation on a class and a method, then the method one will be taken into account. I would have to poke around the spec to verify that.

Reply:
> At the moment this is not really possible with CDI but it would be great if we could support it somehow.

@stuartwdouglas @michalszynkiewicz an interceptor can inject any bean so if we could collect/serialize the info (roles allowed) during build then we could inject/use it at runtime.

Reply:
@mkouba you mean collect them and expose as a bean? That sounds interesting.

Reply:
I chatted with @mkouba and @manovotn about it: while it's doable to prepare the data for the check in the build time, it's not that straighforward and the perf impact of doing it as it is is not big.
For this PR I'd rather keep it in the runtime (it's already quite big now) and, if needed, create an issue to move it to build time later on/in a separate PR

Reply:
I created https://github.com/quarkusio/quarkus/issues/4695 to dicuss it further/implement it later